src/sentry/tasks/app_store_connect.py

```diff
@@ -39,44 +40,70 @@ def inner_dsym_download(
     project = Project.objects.get(pk=project_id)
     config = appconnect.AppStoreConnectConfig.from_project_config(project, config_id)
     client = appconnect.AppConnectClient.from_config(config)
-    itunes_client = client.itunes_client()
-
-    builds = client.list_builds()
-    for build in builds:
-        try:
-            build_state = AppConnectBuild.objects.get(
-                project=project,
-                app_id=build.app_id,
-                platform=build.platform,
-                bundle_short_version=build.version,
-                bundle_version=build.build_number,
-            )
-        except AppConnectBuild.DoesNotExist:
-            build_state = AppConnectBuild(
-                project=project,
-                app_id=build.app_id,
-                bundle_id=config.bundleId,
-                platform=build.platform,
-                bundle_short_version=build.version,
-                bundle_version=build.build_number,
-                fetched=False,
-            )
 
+    # persist all fetched builds into the database as "pending"
+    builds = []
+    for build in client.list_builds():
+        build_state = get_or_create_persisted_build(project, config, build)
         if not build_state.fetched:
-            with tempfile.NamedTemporaryFile() as dsyms_zip:
-                try:
-                    itunes_client.download_dsyms(build, pathlib.Path(dsyms_zip.name))
-                except appconnect.NoDsymsError:
-                    logger.debug("No dSYMs for build %s", build)
-                    continue
+            builds.append(build_state)
+
+    itunes_client = client.itunes_client()
+    for build_state in builds:
+        with tempfile.NamedTemporaryFile() as dsyms_zip:
+            try:
+                itunes_client.download_dsyms(build, pathlib.Path(dsyms_zip.name))
+
                 create_difs_from_dsyms_zip(dsyms_zip.name, project)
-            build_state.fetched = True
-            build_state.save()
-            logger.debug("Uploaded dSYMs for build %s", build)
+                logger.debug("Uploaded dSYMs for build %s", build)
```

Comment:
I would not leave these two lines inside **this** try block.  It is always good to keep the try block as targeted as possible and here you're targeting exceptions from the download, so only put the download in the block.

It could make sense to foresee issues in the parsing of the zips/difs and not fail all builds because one can not be processed.  But that needs it's dedicated try block (though I don't mind if that's not done in this PR)

Reply:
Not sure I follow. This is completely intentional. I want to skip these two lines when the download fails, but *still* mark the build_state as fetched, so I can’t use `continue` in the error case.

Reply:
you can move `create_difs_from_dsyms_zip()` (and the log message) to an `else` clause of this try statement and then we have all the desired behaviour.
```
try:
    itunes_client.download_dsyms(...)
except NoDsymsError:
    ...
except ITunesSessionExpiredException:
   ...
else:
    create_difs_from_dsyms_zip(...)
    logger.debug(...)
```
(I tried to do a suggestion but github is confused and doesn't let me)